stan/math/rev/mat/functor/kinsol_data.hpp

```diff
@@ -0,0 +1,146 @@
+class kinsol_system_data {
+  const F1& f_;
+  const F2& J_f_;
+  const Eigen::VectorXd& x_;
+  const Eigen::VectorXd& y_;
+  const size_t N_;
+  const std::vector<double>& dat_;
+  const std::vector<int>& dat_int_;
+  std::ostream* msgs_;
+
+  typedef kinsol_system_data<F1, F2> system_data;
+
+ public:
+  N_Vector nv_x_;
+  SUNMatrix J_;
+  SUNLinearSolver LS_;
+
+  /* Constructor */
+  kinsol_system_data(const F1& f, const F2& J_f, const Eigen::VectorXd& x,
+                     const Eigen::VectorXd& y, const std::vector<double>& dat,
+                     const std::vector<int>& dat_int, std::ostream* msgs)
+      : f_(f),
+        J_f_(J_f),
+        x_(x),
+        y_(y),
+        dat_(dat),
+        dat_int_(dat_int),
+        msgs_(msgs),
+        N_(x.size()),
+        nv_x_(N_VMake_Serial(N_, &to_array_1d(x_)[0])),
+        J_(SUNDenseMatrix(N_, N_)),
+        LS_(SUNLinSol_Dense(nv_x_, J_)) {}
+
+  ~kinsol_system_data() {
+    N_VDestroy_Serial(nv_x_);
+    SUNLinSolFree(LS_);
+    SUNMatDestroy(J_);
+  }
+
+  /* Implements the user-defined function passed to KINSOL. */
+  static int kinsol_f_system(N_Vector x, N_Vector f, void* user_data) {
+    const system_data* explicit_system
+        = static_cast<const system_data*>(user_data);
+    explicit_system->f_system(NV_DATA_S(x), NV_DATA_S(f));
+
+    return 0;
+  }
+
+  /**
+   * Implements the function of type CVDlsJacFn which is the user-defined
+   * callbacks for KINSOL to calculate the jacobian of the system.
+   * The Jacobian is stored in column major format.
+   *
+   * REMARK - tmp1 and tmp2 are pointers to memory allocated for variables
+   * of type N_Vector which can be used by KINJacFN (the function which
+   * computes the Jacobian) as temporary storage or work space.
+   * See
+   * https://computation.llnl.gov/sites/default/files/public/kin_guide-dev.pdf,
+   * page 55.
+   */
+  static int kinsol_jacobian(N_Vector x, N_Vector f, SUNMatrix J,
+                             void* user_data, N_Vector tmp1, N_Vector tmp2) {
+    const system_data* explicit_system
+        = static_cast<const system_data*>(user_data);
+    return explicit_system->jacobian_states(NV_DATA_S(x), J);
+  }
+
+ private:
+  /**
+   * Calculates the root function, using the user-supplied functor
+   * for a given value of x.
+   */
+  inline void f_system(const double x[], double f[]) const {
+    const std::vector<double> x_vec(x, x + N_);
+    const std::vector<double>& f_vec
+        = to_array_1d(f_(to_vector(x_vec), y_, dat_, dat_int_, msgs_));
+    std::move(f_vec.begin(), f_vec.end(), f);
+  }
+
+  /**
+   * Calculate the Jacobian of the system function with respect to x
+   * using the method specified by J_f_.
+   * By default, J_f is constructed as a method that computes the
+   * Jacobian with reverse-mode autodiff.
+   */
+  inline int jacobian_states(const double x[], SUNMatrix J) const {
+    return J_f_(f_, x_, y_, dat_, dat_int_, msgs_, x, J);
+  }
```

Comment:
Ok, I see. What you have in mind is:
```
return
      static_cast<const system_data*>(user_data)->
      J_f_(f_, x_, y_, dat_, dat_int_, msgs_, NV_DATA_S(x), J);
```
which currently generates some compilation errors
```
error: invalid use of member 'f_' in static member function
```
and so forth for `x_`, `y_`, etc. Now further down the line, Kinsol requires `kinsol_jacobian` to return a `static_int`. I'll admit static variables are not my expertise, but my inclination is not to pass `f_`, `x_`, etc as static variables. Maybe you have a more informed opinion.

Reply:
From the previous comment, after `user_data` cast to `explicit_system`, you can use it to access object members, accessible through `->`. So instead of defining `jacobian_states`, you can just do
`return explicit_system -> J_f_(explicit_system -> f_, explicit_system -> x_, explicit_system -> y_, ...)` to get Jacobian.

Reply:
Except `explicit_system` is not defined (after we remove the extra layer constructing it). So you would have
```
    return
      static_cast<const system_data*>(user_data)->
         J_f_(static_cast<const system_data*>(user_data)->f_,
              static_cast<const system_data*>(user_data)->x_,
              static_cast<const system_data*>(user_data)->y_,
              static_cast<const system_data*>(user_data)->dat_,
              static_cast<const system_data*>(user_data)->dat_int_,
              static_cast<const system_data*>(user_data)->msgs_,
              NV_DATA_S(x), J);
```
which works.

Reply:
`jacobian_states` is called in `kinsol_jacobian`, where `explicit_system` is created? I don't understand what you mean by `remove the extra layer constructing it`. The extra layer in my comment is function `jacobian_states`. What you can do is
```c++
  static int kinsol_jacobian(N_Vector x, N_Vector f, SUNMatrix J,
                             void* user_data, N_Vector tmp1, N_Vector tmp2) {
    const system_data* alg;
        = static_cast<const system_data*>(user_data);
    return alg->J_f_(alg -> f_, alg -> x_, alg -> y_, ..., NV_DATA_S(x), J);
```
Same for `f_system`.